### DISCUSSION TITLE
How to import a path with npm specifiers?

### DISCUSSION THREAD

**Discussion by outloudvi** [denoland/deno#17004] [Q&A] [answered]
It's apparently possible to achieve this:

``` ts
import uniq from "https://cdn.skypack.dev/lodash@4.17.21/uniq"
console.log(uniq([1,2,2,3])) // [1,2,3]
```

With the [npm specifiers](https://deno.land/manual@v1.28.3/node/npm_specifiers), it's also possible to import a whole package from npm:

``` ts
import uniq from "npm:lodash.uniq@4.5.0"
console.log(uniq([1,2,2,3])) // [1,2,3]
```

Do we have a way to achieve:

``` ts
// It currently doesn't work for me
import uniq from "npm:lodash@4.17.21/uniq"
console.log(uniq([1,2,2,3])) // [1,2,3]
```

**Comment by bartlomieju** [comment, marked as answer]
You should be able to use import map like so:
```json
{
    "imports": {
       "lodash/": "npm:/lodash@4.17.21"
    }
}
```
```js
import uniq from "lodash/uniq";
```

This was fixed in https://github.com/denoland/deno/pull/16754, notice the slash after `npm:` - it's required in import maps to map to a folder.

**Reply by outloudvi** [reply]
Thanks! Seems that it works, but only with a `.js` suffix on the import:

``` ts
import uniq from "lodash/uniq.js"
console.log(uniq([1,2,2,3])) // [1,2,3]
```

It also asked me to add a trailing slash at `import_map.json`:
``` json
{
    "imports": {
       "lodash/": "npm:/lodash@4.17.21/"
    }
}
```

Then it can give me:
```
$ deno run --importmap=import_map.json ./test.ts
[ 1, 2, 3 ]
```

**Reply by Conaclos** [reply]
Is it the recommended way to import a _npm_ module as a library author? If I understand correctly, importing a dependency in _deno_ does not take the dependency's import map into account?

**Reply by bartlomieju** [reply]
> Thanks! Seems that it works, but only with a `.js` suffix on the import:
> 
> ```ts
> import uniq from "lodash/uniq.js"
> console.log(uniq([1,2,2,3])) // [1,2,3]
> ```
> 
> It also asked me to add a trailing slash at `import_map.json`:
> 
> ```json
> {
>     "imports": {
>        "lodash/": "npm:/lodash@4.17.21/"
>     }
> }
> ```
> 
> Then it can give me:
> 
> ```
> $ deno run --importmap=import_map.json ./test.ts
> [ 1, 2, 3 ]
> ```

Ah yes, if these are ES modules then the `.js` extension is indeed required.

**Reply by bartlomieju** [reply]
> Is it the recommended way to import a _npm_ module as a library author? If I understand correctly, importing a dependency in _deno_ does not take the dependency's import map into account?

That's correct, library authors are not meant to use import maps - they do not "cascade" or "stack", only the final import map is applied.

**Reply by Conaclos** [reply]
> That's correct, library authors are not meant to use import maps - they do not "cascade" or "stack", only the final import map is applied.

Is there any plan to make it cascading?

**Reply by bartlomieju** [reply]
Not unless the WICG reaches a consensus on this feature. See issues on https://github.com/WICG/import-maps

**Comment by spencerhakim** [comment]
I think your example should work as-is, #15578 should have added support for this back in August. If it isn't working, you might want to file a bug report.

**Reply by outloudvi** [reply]
Thanks for your pointing to that PR! I see that precisely described my need, but probably something changed over the later commits.
